ioos_qc/qartod.py

```diff
@@ -387,11 +385,14 @@ def check(self, tinp, inp, zinp):
                 fail_idx = np.zeros(inp.size, dtype=bool)
 
             suspect_idx = (inp < m.vspan.minv) | (inp > m.vspan.maxv)
-
+            
             with np.errstate(invalid='ignore'):
                 flag_arr[(values_idx & fail_idx)] = QartodFlags.FAIL
                 flag_arr[(values_idx & ~fail_idx & suspect_idx)] = QartodFlags.SUSPECT
                 flag_arr[(values_idx & ~fail_idx & ~suspect_idx)] = QartodFlags.GOOD
```

Comment:
The only thing is.. I am curious why the missing value flag of 9 gets overridden here and turns into a 1. When you print the masked array at this step the missing values are still masked but for some reason `flag_arr[(values_idx & ~fail_idx & ~suspect_idx)]` includes the masked value as well as the good value. And assigns both of them with a 1.

Reply:
@iwensu0313 I won't pretend that my jet-lagged brain can follow all the changes but they LGTM and looks like a good improvement from the current state.

Do you want to work a bit more on this or are you good should we merge?

Reply:
Thanks @ocefpaf - I'll add more details to my summary. Also I'm double checking the above described behavior with the team/@kthyng. Because it might be related to dependencies (e.g. numpy version). Once I resolve this thread I am happy to merge

Reply:
Thanks for merging @ocefpaf. I dug into this masked array behavior a bit but it is really absurd and don't totally know why it's happening. Documenting here in case we need to revisit in the future. but I am going to resolve this thread bc I've run out of ideas and don't have time right now.

I ran through the scenario lgarzio provided in #65 that does work (setting `zspan` to `(0,1000)` in `QartodClimatologyMissingTest`) and the only difference is that `values_idx` is a masked array instead of a normal array. 

So I changed the way we define `z_idx`
```
z_idx = np.ma.array(data=np.ones(inp.size, dtype=bool), mask=inp.mask, fill_value=999999)
```
to mimic what `z_idx` looks like in the scenario that works as expected. But we still get the same output.
```
> z_idx
masked_array(data=[True, --, True],
             mask=[False,  True, False],
       fill_value=True)
```
Screenshots for reference.

<img width="414" alt="Screenshot 2024-04-02 at 8 40 49 AM" src="https://github.com/ioos/ioos_qc/assets/9053844/69769ce5-bd0e-4e9e-81bf-ec9b846239c8">

<img width="394" alt="Screenshot 2024-04-02 at 8 38 27 AM" src="https://github.com/ioos/ioos_qc/assets/9053844/5919b308-186b-47bb-bab7-d3c3d09667e9">

Changed `z_idx` to be a masked array to match 
<img width="1062" alt="Screenshot 2024-04-02 at 9 17 37 AM" src="https://github.com/ioos/ioos_qc/assets/9053844/9d293d9e-55bd-4922-83ff-c0555804b7de">

Reply:
@kthyng figured it out!! We need to make an update and replace `z_idx = np.ones(inp.size, dtype=bool)` with `z_idx = np.ma.array(data=~np.isnan(inp.data), mask=inp.mask, fill_value=999999)` so that the missing value flag doesn't get overwritten with the good data flag. I'd even move the `flag_arr[(values_idx & ~fail_idx & ~suspect_idx)] = QartodFlags.GOOD` back to the top of `check()` to ensure that it continues to work with changing `numpy` package versions. Because my fix was sort of a bandaid solution.

In plain English, the fix Kristen figured out ensures that `z_idx` is created w/ `inp.data` which preserves the underlying information about what data points are masked values. Without it, python doesn't know it's masked and that's why it is kept when we slice it here: `flag_arr[(values_idx & ~fail_idx & ~suspect_idx)] = QartodFlags.GOOD`

Reply:
Sorry for merging this before you were done. I wanted to make it easier for others to test. We can finalize this today and mint a new release. Thanks you both for figuring this out!